### PR TITLE
Allow custom payment metadata inside SDK storage

### LINKED ISSUE #635: Feature request: Allow SDK users to store additional metadata inside SDK storage
It would be useful for SDK users to be able to persist additional metadata inside the SDK storage.

Some metadata might be **associated with payments** (e.g. fiat exchange rate, timezone, user category).
Also it would be nice if:
* the SDK allows to search by such data using `list_payments()`
* SDK encrypts the data (since such data might be private)
* the SDK sync the data between different instances

Probably a solution might be a field allowing to store a JSON-like structure or multiple fields storing key-value pairs.

Another usecase is to persist data **not associated with a particular payment** like users settings (e.g. fiat currency, language).

### PR DESCRIPTION
**Pull request by hydra-yse**
Closes #635

This PR aims to address modifying metadata strictly related to payments. A similar feature is already present in the persister:
https://github.com/breez/breez-sdk/blob/e6323f57acf0a5aad58974695b60ce07b0b77bfe/libs/sdk-core/src/persist/transactions.rs#L69-L74

Yet the method expects a pre-defined struct. This could be changed to accept any kind of JSON-seralized string, as Sqlite also [supports JSON operators and fields](https://www.sqlite.org/json1.html).  This would be the cleaner approach, rather than creating a separate method with equivalent functionality, but it has the following caveats:

- Currently the `insert_payment_external_info` method is also used internally for metadata such as lnurl-related fields, or errors. Merging the two into a single metadata column could lead to the possibility of developers overwriting these values, if they do not call the proper `update_payment_external_info` method (the two methods are split due to sqlite's technical limitations, if "insert" is called, the whole column will be replaced with the specified object). This is not really clean and prone to error for the end developer, and I am looking forwards to possible alternatives (possibly not involving an additional database read).

**Edit:** ~To address the first caveat, a SQL trigger can be created which automatically generates an empty JSON object whenever a payment is added to the `payments` table. Thus, one can simply call insert and update that row with the corresponding fields.~ This does not seem to be possible as the two tables belong to different databases. The best solution does indeed seem to perform an additional read to check whether the row exists already.

### CODE REVIEW

**Comment by hydra-yse** [comment]
Based on your comments, I opted for keeping the struct and adding an `external_metadata` column instead. Let me know if the name sounds repetitive/needs change. I also added some utilities like update/retrieve functions that coerce directly into the desired struct.

File: libs/sdk-core/src/persist/migrations.rs
Comment (dangeross) [inline libs/sdk-core/src/persist/migrations.rs:183]: We cannot insert or edit migrations defined already in the `current_migrations` / `current_sync_migrations`, as these migrations have already been run on existing clients using the SDK. Any migration needed has to be added to the end of the respective list.

So this line needs to be moved to the end of the `current_sync_migrations` list of migrations and you can remove the other changes here.

**Review by dangeross — CHANGES_REQUESTED** [review]

File: libs/sdk-core/src/persist/migrations.rs
Comment (hydra-yse) [reply libs/sdk-core/src/persist/migrations.rs:183]: Understood, thanks Ross

File: libs/sdk-core/src/breez_services.rs
Comment (roeierez) [inline libs/sdk-core/src/breez_services.rs:580]: I think we need to add a "metadata" field to the payment structure (returned by list_payments). Otherwise the user would have to iterate over the list and fetch one by one.
If we do so the advantage is that we can remove this method as we already have `payment_by_hash`.

File: libs/sdk-core/src/breez_services.rs
Comment (roeierez) [inline libs/sdk-core/src/breez_services.rs:588]: NIT: Perhaps we can shorten this to `set_payment_metadata` ?
Also missing the udl file implementation (In case you wanted to hold this until the approach is approved then ignore this comment).

**Review by roeierez — CHANGES_REQUESTED** [review]
Looks nice and simple. Added come comment to simplify a bit more.

File: libs/sdk-core/src/persist/sync.rs
Comment (hydra-yse) [inline libs/sdk-core/src/persist/sync.rs:174]: As a side note, I was wondering whether or not it would would be useful to keep the `external_metadata` in the backup, as its usefulness could vary based on what the implementation is like. Given the small size, perhaps it is fine for now?

Comment (roeierez) [reply libs/sdk-core/src/persist/sync.rs:174]: I agree.  It only I think it is fine but also it seems mandatory since we want this metadata to sync accross apps/devices

File: libs/sdk-core/src/breez_services.rs
Comment (roeierez) [inline libs/sdk-core/src/breez_services.rs:578]: If we require the string to be json then we should validate it is a valid json otherwise we will end up with invalid json in the db.
Also I have noticed int the original issue it was mentioned that it would be nice to be able to search payments for a specific metadata value. This can be done using JSON1 sqlite extension (https://sqlite.org/json1.html). I think we can defer this specific search requirement to a following PR but we need to verify our current solution will be able to supports it.

File: libs/sdk-core/src/persist/sync.rs
Comment (roeierez) [inline libs/sdk-core/src/persist/sync.rs:176]: Syncing metadata is a bit different since we allow updating it in contrary to other columns that are created once and immutable.
This creates a problematic scenario when one device sets the metadata and after sync to a remote one subsequent updates won't be synced again (since the payment already exists).
One way to solve this is use the updated_at timestamp to sync the same way we do with the swap fees (`sync_swaps_fees_local`)

**Review by roeierez — CHANGES_REQUESTED** [review]

File: libs/sdk-core/src/breez_services.rs
Comment (hydra-yse) [reply libs/sdk-core/src/breez_services.rs:578]: You're right about verifying the JSON. I read that SQLite's `json_set` did checks before inserting, but I just tested out that it is not the case. Regarding searching for specific fields, I believe we can support it right away by exposing a method using `json_extract`.

**Review by dangeross — COMMENTED** [review]
LGTM so far, just need to resolve the JSON validation and sync updating that Roei mentioned.

File: libs/sdk-core/src/models.rs
Comment (hydra-yse) [inline libs/sdk-core/src/models.rs:551]: Just wanted to leave two quick notes regarding the enum:
1. The choice to have the values specified in such a way (using a struct with `value` field) is due to UniFFI's limitations in regards to enums.
2. When `IsNull` is specified, due to limitations in SQLite's json methods when a field does not exist in an object, it will be treated as if it were null (e.g. in the following example, querying for "field" IsNull will return both objects): 
```js
  obj1 = {
     "field": null  
  }
  obj2 = {}
  ```
This is not a problem in most scenarios, especially when all the payment metadatas retain the same structure, but I believe it should be properly documented anyway.

File: libs/sdk-bindings/src/breez_sdk.udl
Comment (roeierez) [inline libs/sdk-bindings/src/breez_sdk.udl:209]: Perhaps we can simplify this and use strings only instead of JsonFieldValue ?

Comment (roeierez) [reply libs/sdk-bindings/src/breez_sdk.udl:209]: Also we should think about this interface a bit more as the sent/received/closed channels are used as union in the query and metadata is used as filter.

Comment (hydra-yse) [reply libs/sdk-bindings/src/breez_sdk.udl]: Addressed both points in the following commit. Would've been nice to use a `HashMap<String, String>` instead of a `Vec<PaymentMetadata>` for the filters, but apparently flutter_rust_bridge does not seem to support them (https://github.com/fzyzcjy/flutter_rust_bridge/issues/329).

File: libs/sdk-core/src/persist/sync.rs
Comment (roeierez) [inline libs/sdk-core/src/persist/sync.rs:178]: The problem with this approach is that if one device updates the metadata and wasn't synced other devices will override their data on next sync (since they would update the whole item)
Perhaps it is better to use a different table for metadata?

File: libs/sdk-bindings/src/breez_sdk.udl
Comment (roeierez) [inline libs/sdk-bindings/src/breez_sdk.udl:202]: NIT: rename to PaymentMetadataFilter

File: libs/sdk-core/src/breez_services.rs
Comment (roeierez) [inline libs/sdk-core/src/breez_services.rs:578]: I still don't see json validation and even if we do so what happen if the user uses a nested json? Do we allow this? Won't it break the search?
How about passing here an array of metadata where every metadata has `key` and `value` fields where both are strings? We can validate for duplicate keys and it makes the search a lot simpler since we only use strings as values.
@dangeross what do you think?

**Review by roeierez — CHANGES_REQUESTED** [review]

File: libs/sdk-core/src/breez_services.rs
Comment (hydra-yse) [reply libs/sdk-core/src/breez_services.rs:578]: I added the JSON validation inside the `update_payment_external_metadata` method, see [here](https://github.com/hydra-yse/breez-sdk/blob/febab11e462fe20d8f334eaf5e178711150bcc68/libs/sdk-core/src/persist/transactions.rs#L110-L111). Regarding passing a vector of `PaymentMetadata`, I agree, since we can then also reduce everything to one SQL query.

File: libs/sdk-bindings/src/breez_sdk.udl
Comment (hydra-yse) [reply libs/sdk-bindings/src/breez_sdk.udl:202]: Given the other reply about passing a vector to `set_payment_metadata`, maybe we should keep this as PaymentMetadata and use it for both filtering and inserting. What do you think?

Comment (roeierez) [reply libs/sdk-bindings/src/breez_sdk.udl:202]: The filter structure might change in the future ( we may for example add support for operators (contains, equals, etc...).
I think it is better to have a separate structure even though currently they look the same, they represent different things.

File: libs/sdk-core/src/breez_services.rs
Comment (dangeross) [reply libs/sdk-core/src/breez_services.rs:578]: As @hydra-yse says, you can query within the nested JSON using `json_extract` with a key like `$.parent.child`, even with arrays `$.parent.arr[*].child`. I don't have a problem with the metadata being JSON, I think the question is what is the best approach to use to sync the metadata. A key/value store in a separate table is probably a better approach to keep the payment syncing clean. With a separate table when syncing we can control the approach to updating we want to take based on the metadata payment_hash/key, be it write once or write many (based on timestamp).

Comment (roeierez) [reply libs/sdk-core/src/breez_services.rs:578]: I agree a separate table is the correct approach as the sync strategy is different.
My initial thinking was one row per payment with a json field.
When you referring to key/value store do you mean one row per metadata key so there is one to many relation between payment to metadata? This will complicate the interface as we will need to support also delete.

Comment (dangeross) [reply libs/sdk-core/src/breez_services.rs:578]: Ah yeah, I was thinking one to many with one row per key/value but didn't consider the delete case. Then one row per payment_hash/json seems more appropriate.

Comment (ok300) [reply libs/sdk-core/src/breez_services.rs:578]: With JSON, the question of validation remains. AFAIK duplicate keys are valid in JSON, but would be invalid in our scenario. So we'd have to explicitly check and deny that.

File: libs/sdk-core/src/persist/transactions.rs
Comment (hydra-yse) [inline libs/sdk-core/src/persist/transactions.rs:414]: Hope you like this approach. After our conversation, I first attempted implementing strings-only, yet realized that if we pass a JSON object for filtering as well (with the desired JSONPath-value fields) I can pretty easily coerce them into the right type! This should support the majority of cases, and is pretty intuitive for the end developer.

File: libs/sdk-core/src/persist/migrations.rs
Comment (dangeross) [inline libs/sdk-core/src/persist/migrations.rs:425]: I think the table only needs creating in the sync storage

File: libs/sdk-core/src/persist/transactions.rs
Comment (dangeross) [inline libs/sdk-core/src/persist/transactions.rs:117]: Could replace this with the ensure_sdk! macro
```suggestion
        ensure_sdk!(
            new_metadata.len() <= METADATA_MAX_SIZE_KB,
            PersistError::Generic(
                anyhow!("Max metadata size ({}MB) has been exceeded", METADATA_MAX_SIZE_KB / 1000)
            )
        );         
```

File: libs/sdk-core/src/persist/transactions.rs
Comment (dangeross) [reply libs/sdk-core/src/persist/transactions.rs:414]: Sounds good. So something like?
```
{ "path.to.field": true }
```

File: libs/sdk-core/src/models.rs
Comment (dangeross) [inline libs/sdk-core/src/models.rs:644]: I'm not sure if filter or filters is better here. I lean towards metadata_filter because its not a list.
```suggestion
    pub metadata_filter: Option<String>,
```

**Review by dangeross — CHANGES_REQUESTED** [review]
Looking pretty good!

File: libs/sdk-core/src/persist/transactions.rs
Comment (hydra-yse) [reply libs/sdk-core/src/persist/transactions.rs:13]: No, it's actually kilobytes. Now that I think about it though, it probably is too much. Let's set a max of 10kb, and see how it goes, since the string encoding is UTF-8 and may be heavier for some locales (up to 4 bytes per character, meaning 2500 characters)

File: libs/sdk-core/src/persist/transactions.rs
Comment (hydra-yse) [reply libs/sdk-core/src/persist/transactions.rs:117]: Didn't know about it. Thanks Ross!

File: libs/sdk-core/src/persist/transactions.rs
Comment (dangeross) [reply libs/sdk-core/src/persist/transactions.rs:13]: 👍 I thought 10MB might be too much 😄

**Comment by yaslama** [comment]
Imo, the best way is to use the -> operator and pass 2 strings to the filter: the first one is a json path, the second one a json representation
You can try for instance to run the following (command line sqlite3 or https://sqlite.org/fiddle/):
```
CREATE TABLE t(a json,b int);
INSERT INTO t(a,b) VALUES('{"a":2,"c":[4,5,{"f":7}]}',123),('[11,22,33,44]',456),(NULL,789),('"aa"',012);
INSERT INTO t(a,b) VALUES('{"a":true,"c":[4,5,{"f":7}]}',789);
INSERT INTO t(a,b) VALUES('{"a":1,"c":[4,5,{"f":7}]}',987);

select * from t where a->'$.c[2]'='{"f":7}';
select * from t where a->'$.c[2].f'='7';
select * from t where a->'$.a'='1';
select * from t where a->'$.a'='true';
```

File: libs/sdk-core/src/persist/transactions.rs
Comment (hydra-yse) [inline libs/sdk-core/src/persist/transactions.rs:406]: Thanks @yaslama for the reply. This does indeed simplify even more our approach, although I noticed we still need to wrap JSON strings inside some double quotes in order to make this work as intended. Still, a huge improvement over checking for null and boolean fields, as now it can be done with only one query 👍

Comment (yaslama) [reply libs/sdk-core/src/persist/transactions.rs:406]: > although I noticed we still need to wrap JSON strings inside some double quotes in order to make this work as intended

I don't understand this sentence. In order to be a valid json, a string has to be enclosed by double quotes. So if the second parameter (the value of the filter) is required to be a json, why it's a problem?

Comment (hydra-yse) [reply libs/sdk-core/src/persist/transactions.rs:406]: Since we need to parse the filter string (a JSON object) and split it into multiple where-clauses for each field, we need to iterate over a `Map<String, Value>`. When a field has a `Value::String`, and we call the method `to_string()` (used in `format!` macro), the string gets printed out without the double quotes, which is a problem for the SQL query as it only works as follows: `SELECT * FROM metadata->'$.{fieldName}' = '"{value}"'`

Comment (roeierez) [reply libs/sdk-core/src/persist/transactions.rs:406]: @hydra-yse I think What Akiba's suggestion is how to solve the issue of searching in non string values.
And by solving this it allows us to go back to the previously suggested interface for filtering instead of passing a json string as filter.

We can pass a list of `PaymentMetadataFilter` where the struct is defined like this:
```
struct PaymentMetadataFilter {
  search_path: String,
  search_value: String
}
```

Then it is trivial to compose the where clause using the arrow operator and we get two benefits:
1. The interface is clear (we pass a list of filters with documented fields)
2. It will search in all values (including booleans, numbers)

What do you think?

Comment (hydra-yse) [reply libs/sdk-core/src/persist/transactions.rs:406]: Sure, they would both work well. On one hand I like the simplicity of being able to pass a JSON string without requiring any additional data structure, but I also see how setting the filters this way would make it very similar (and thus more familiar) to the `type_filters`. I'll continue as you suggested

File: libs/sdk-core/src/models.rs
Comment (roeierez) [inline libs/sdk-core/src/models.rs:558]: I don't understand the comment here. Should the user add quotes as part of the value here?

File: libs/sdk-core/src/persist/transactions.rs
Comment (roeierez) [inline libs/sdk-core/src/persist/transactions.rs:13]: I would start with 1k, we can always expand later.

File: libs/sdk-core/src/persist/transactions.rs
Comment (roeierez) [inline libs/sdk-core/src/persist/transactions.rs:403]: Since we are formatting here a string using external input we should validate the input for security reasons.
Best would be to use a prepared statement and pass these values as parameters (at least the value).
We can validate the json path using https://docs.rs/serde_json_path/latest/serde_json_path/struct.JsonPath.html

**Review by roeierez — CHANGES_REQUESTED** [review]
Looks good. I think we are very close.

File: libs/sdk-core/src/models.rs
Comment (hydra-yse) [reply libs/sdk-core/src/models.rs:558]: Yes, if the value inside the JSON object is a JSON string, the developer must add double quotes in order to match for it. The `->` operator simply creates a one-to-one match, so "string" can only be matched by the literal "string".

Comment (roeierez) [reply libs/sdk-core/src/models.rs:558]: I see, I think it will create a confusion. Users will not be aware of that, they will just pass a string...
@yaslama any idea on how to make the experience here more straight forward?

Comment (roeierez) [reply libs/sdk-core/src/models.rs:558]: @hydra-yse @yaslama perhaps we can trim the quotes before the matching?

`trim(metadata->'$.{}', '"') = '{}'`

File: libs/sdk-core/src/persist/transactions.rs
Comment (hydra-yse) [reply libs/sdk-core/src/persist/transactions.rs:403]: I agree using a prepared statement, because that is enough to prevent SQL injections, but is parsing the JSON path necessary? The worst that can happen if someone specifies a bad path would just be a bad filter that does not catch anything, which is not an issue on our side.

What do you think?

Comment (roeierez) [reply libs/sdk-core/src/persist/transactions.rs:403]: I think If we use a prepared statement then we don't need to parse the json path.

File: libs/sdk-core/src/models.rs
Comment (roeierez) [reply libs/sdk-core/src/models.rs:558]: @hydra-yse  Now thinking again my suggestion will make a problem in case the user pass json (nested).
Ok, I am fine with the quotes as we document the value should be a valid json (where strings should be quoted)

File: libs/sdk-core/src/persist/transactions.rs
Comment (dangeross) [inline libs/sdk-core/src/persist/transactions.rs:408]: I wonder if these single quotes are the issue:
```suggestion
            where_clause.push("metadata->?=?".to_string());
```
then when building the params:
```
metadata_filters.iter().for_each(|PaymentMetadataFilter { search_path, search_value }| {
    params.push(format!("$.{}", search_path));
    params.push(search_value.clone());
})
```

File: libs/sdk-core/src/persist/transactions.rs
Comment (hydra-yse) [inline libs/sdk-core/src/persist/transactions.rs:248]: Based on the solution proposed by @roeierez and @dangeross , I decided to opt for their approach. This still does not work, but at the same time the parser does not complain of any parsing errors. I am thinking the issue is now related to the JSON Path, as maybe it is not evaluated as such due to the fact that the `$` is part of the prepared statement, perhaps? I have verified the values get indeed passed to `query_map` as intended, therefore I imagine it to be a SQL-related issue.

Comment (roeierez) [reply libs/sdk-core/src/persist/transactions.rs:248]: The test you added did pass using the commit I posted: https://github.com/breez/breez-sdk/commit/7a324252a22e2954ad93cb83e55890b1c836cdbc

Comment (hydra-yse) [reply libs/sdk-core/src/persist/transactions.rs:248]: @roeierez Thanks for the swift reply, turns out the quotes were not needed at all. Made it work with the HashMap, pushing rn!

File: libs/sdk-bindings/src/breez_sdk.udl
Comment (roeierez) [inline libs/sdk-bindings/src/breez_sdk.udl:203]: NIT: perhaps json_path to be more explicit?

File: libs/sdk-bindings/src/breez_sdk.udl
Comment (roeierez) [inline libs/sdk-bindings/src/breez_sdk.udl:204]: Same here, `json_value` ?

File: libs/sdk-bindings/src/breez_sdk.udl
Comment (roeierez) [inline libs/sdk-bindings/src/breez_sdk.udl:202]: NIT: rename to `MetadataFilter` ? This looks generic enough to apply (if needed) later on other entities.

**Review by roeierez — APPROVED** [review]
LGTM! I added several NIT suggestions about renaming fields, other than that good to go IMO.

**Review by dangeross — APPROVED** [review]
Looks good @hydra-yse, just one small typo

**Pull request merged**
